sktime/annotation/e_agglo.py

```diff
@@ -0,0 +1,427 @@
+# -*- coding: utf-8 -*-
+"""E-agglo."""
+
+import warnings
+
+import numpy as np
+import pandas as pd
+
+from sktime.transformations.base import BaseTransformer
+
+__author__ = ["KatieBuc"]
+__all__ = ["EAGGLO"]
+
+
+class EAGGLO(BaseTransformer):
+    """
+    Docstring.
+
+    Parameters
+    ----------
+    ...
+
+    Attributes
+    ----------
+    ...
+
+    References
+    ----------
+    .. [1] ...
+
+    Examples
+    --------
+    ...
+    """
+
+    _tags = {
+        "fit_is_empty": False,
+    }
+
+    def __init__(
+        self,
+        member=None,
+        alpha=1.0,
+        penalty=None,
+    ):
+        self.member = member
+        self.alpha = alpha
+        self.penalty = penalty
+        super(EAGGLO, self).__init__()
+
+    def _process_data(self, X):
+        """Docstring."""
+        member_ = (
+            self.member if self.member is not None else np.array(range(X.shape[0]))
+        )
+
+        u = np.sort(np.unique(member_))  # unique array of cluster labels
+        N_ = len(u)  # number of clusters
+
+        for i in range(
+            N_
+        ):  # relabel clusters to be consecutive numbers (when user specified)
+            member_[np.where(member_ == u[i])[0]] = i
+
+        # check if sorted.
+
+        sizes_ = np.repeat(0, 2 * N_)
+        sizes_[:N_] = [
+            sum(member_ == i) for i in range(N_)
+        ]  # calculate initial cluster sizes_
+
+        # array of within distances
+        within = [
+            get_within(
+                X.loc[
+                    member_ == i,
+                ],
+                self.alpha,
+            )
+            for i in range(N_)
+        ]
+
+        # dataframe of between between-within distances
+        D_ = pd.DataFrame(index=range(2 * N_), columns=range(2 * N_))  # .fillna(np.Inf)
```

Comment:
consider renaming `D_` to sth like `distances`

Reply:
Also, is there an advantage of using a DataFrame here instead of a regular np.array?